no uncertainty restarts for node-local transactions

tbg opened the pull request:
Automatically add the coordinator's `NodeID` to `Txn.CertainNodes`, preventing uncertainty-related restarts from commands addressed to the local node.

Added a logic test verifying that the shell session in #2861 now works, which in turn required setting a non-trivial clock offset for `TestServer` (something we should have done earlier to test closer to production). This demonstrated that the `Lease` subsystem is likely incorrect (#3050).

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3052)

<!-- Reviewable:end -->

tbg committed:
mark own NodeID as certain for new transactions

closes #815, makes progress on #2861

tbg committed:
nontrivial clock offset for TestServer

needed to disable two lease tests, see #3050.

tamird commented:
LGTM

petermattis commented on sql/testdata/snapshot:
```diff
@@ -0,0 +1,58 @@
+# Run a simple shell session in which a reader conflicts with an open txn in
+# SNAPSHOT isolation by means of a read, which is resolved by a Push. Prevents
+# regression of #2861.
```
I noticed in go that they have test cases named after the issue they reproduce (e.g. https://github.com/golang/go/blob/master/test/fixedbugs/issue11326.go). Seems like a nice pattern to follow: `s,testdata/snapshot,testdata,issue2861,g`

tschottdorf replied:
sounds good, I went for `snapshot_issue2861` to add a little more context.

petermattis replied:
What will we do with issues that have no simple context?

tschottdorf replied:
not put a context. The reason I did this is because I doubt I'll ever care
about this being related to that specific issue number, but I'll wind up
looking for tests that have to do with snapshot isolation.

On Mon, Nov 9, 2015 at 10:30 AM, Peter Mattis <notifications@github.com>
wrote:

> In sql/testdata/snapshot
> <https://github.com/cockroachdb/cockroach/pull/3052#discussion_r44287372>:
>
> > @@ -0,0 +1,58 @@
> > +# Run a simple shell session in which a reader conflicts with an open txn in
> > +# SNAPSHOT isolation by means of a read, which is resolved by a Push. Prevents
> > +# regression of #2861.
>
> What will we do with issues that have no simple context?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/3052/files#r44287372>.
>

tbg committed:
verify #2861 is resolved

fixes #2861. With a trivial clock offset (as used prior to the
last commit), this test would pass unconditionally even before
the penultimate commit, but it has been verified that its
behavior has changed.

petermattis commented:
LGTM
